File: java/believe/map/gui/LevelMap.java
package believe.map.gui;

import static believe.geometry.RectangleKt.mutableRectangle;

import believe.core.Updatable;
import believe.core.display.Renderable;
import believe.gui.ComponentBase;
import believe.map.data.BackgroundSceneData;
import believe.map.data.LayerData;
import believe.map.data.MapData;
import believe.scene.GeneratedMapEntityData;
import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.gui.GUIContext;
import javax.annotation.Nullable;
import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@AutoFactory
public class LevelMap extends ComponentBase {
  private final List<MapLayerComponent> rearLayers;
  private final List<MapLayerComponent> frontLayers;
  private final Set<Renderable> renderables;
  private final Set<Updatable> updatables;
  private final List<MapBackground> mapBackgrounds;

  @Nullable private ComponentBase focus;

  @Inject
  public LevelMap(
      @Provided GUIContext container,
      @Provided MapLayerComponentFactory mapLayerComponentFactory,
      MapData mapData) {
    super(container, 0, 0);
    GeneratedMapEntityData.Builder generatedMapEntityDataBuilder =
        GeneratedMapEntityData.newBuilder();
    mapData.tiledMapData().objectLayers().stream()
        .flatMap(layer -> layer.objectFactories().stream())
        .forEach(factory -> factory.createAndAddTo(generatedMapEntityDataBuilder));
    GeneratedMapEntityData generatedMapEntityData = generatedMapEntityDataBuilder.build();

    renderables = generatedMapEntityData.renderables();
    updatables = generatedMapEntityData.updatables();
    rect =
        mutableRectangle(
            rect.getX(),
            rect.getY(),
            mapData.tiledMapData().width(),
            mapData.tiledMapData().height());
    setLocation(0, 0);

    rearLayers = new ArrayList<>();
    frontLayers = new ArrayList<>();
    for (LayerData layer : mapData.tiledMapData().layers()) {
      if (!layer.isVisible()) {
        continue;
      }

      if (layer.isFrontLayer()) {
        frontLayers.add(mapLayerComponentFactory.create(layer));
      } else {
        rearLayers.add(mapLayerComponentFactory.create(layer));
      }
    }
    mapBackgrounds =
        mapData.backgroundScenes().stream()
            .map(
                (BackgroundSceneData backgroundSceneData) ->
                    new MapBackground(backgroundSceneData, mapData.tiledMapData().height()))
            .collect(Collectors.toList());
  }

  public void reset() {
    setLocation(0, 0);
  }

  /**
   * Sets the object that should be centered on screen, like the player
   *
   * @param focus The component to focus on
   */
  public void setFocus(ComponentBase focus) {
    this.focus = focus;
  }

  public void update(int delta) {
    for (Updatable updatable : updatables) {
      updatable.update(delta);
    }
  }

  public List<MapBackground> getBackgrounds() {
    return mapBackgrounds;
  }

  @Override
  public void resetLayout() {}

  @Override
  protected void renderComponent(GUIContext context, Graphics g) throws SlickException {
    for (MapLayerComponent rearLayer : rearLayers) {
      rearLayer.renderComponent(context, g);
    }

    for (Renderable renderable : renderables) {
      renderable.render(context, g);
    }

    if (focus != null) {
      focus.render(context, g);
    }

    for (MapLayerComponent frontLayer : frontLayers) {
      frontLayer.renderComponent(context, g);
    }
  }
}

File: java/believe/scene/LevelMap.kt
package believe.scene

import believe.core.display.Graphics
import believe.geometry.Rectangle
import believe.map.data.MapData
import believe.physics.manager.PhysicsManager
import believe.react.Observer
import dagger.Reusable
import javax.inject.Inject

class LevelMap private constructor(
    val width: Float,
    val height: Float,
    private val mapBackgrounds: List<MapBackground>,
    private val rearLayers: List<LevelMapLayer>,
    private val sceneElements: List<SceneElement>,
    private val frontLayers: List<LevelMapLayer>
) : SceneElement, Observer<Rectangle> {

    override var x: Float = 0f
        set(value) {
            field = value
            rearLayers.forEach { it.x = field }
            frontLayers.forEach { it.x = field }
        }

    override var y: Float = 0f
        set(value) {
            field = value
            rearLayers.forEach { it.y = field }
            frontLayers.forEach { it.y = field }
        }

    override fun render(g: Graphics) {
        mapBackgrounds.forEach { it.render(g) }
        rearLayers.forEach { it.render(g) }
        sceneElements.forEach { it.render(g) }
        frontLayers.forEach { it.render(g) }
    }

    override fun bind() = sceneElements.forEach(SceneElement::bind)
    override fun unbind() = sceneElements.forEach(SceneElement::unbind)
    override fun update(delta: Long) = sceneElements.forEach { it.update(delta) }

    override fun valueChanged(newValue: Rectangle) {
        mapBackgrounds.forEach { it.valueChanged(newValue) }
    }

    @Reusable
    class Factory @Inject internal constructor(private val physicsManager: PhysicsManager) {
        /**
         * Creates a [LevelMap].
         *
         * @param mapData the [MapData] used in rendering the map to the screen.
         * @param extraSceneElements the list of [SceneElement] instances which will be rendered
         * above any other [SceneElement] instances generated from [MapData]. These elements will be
         * rendered in order.
         */
        fun create(mapData: MapData, extraSceneElements: Set<SceneElement>): LevelMap {
            val tiledMapData = mapData.tiledMapData()
            val generatedEntities = GeneratedMapEntityData.newBuilder().apply {
                tiledMapData.objectLayers().flatMap {
                    it.objectFactories()
                }.forEach { it.createAndAddTo(this) }
            }.build()
            val sceneElements = (generatedEntities.sceneElements() + extraSceneElements).toList()

            val layerData =
                tiledMapData.layers().filter { it.isVisible }.partition { it.isFrontLayer }
            generatedEntities.physicsManageables()
                .forEach { it.addToPhysicsManager(physicsManager) }

            val mapHeight = tiledMapData.height()
            return LevelMap(tiledMapData.width().toFloat(),
                mapHeight.toFloat(),
                mapData.backgroundScenes().map { MapBackground(it, mapHeight) },
                layerData.second.map { LevelMapLayer(it.layer()) },
                sceneElements,
                layerData.first.map { LevelMapLayer(it.layer()) })
        }
    }
}